### PR TITLE
added: (Tensor|Vec)Func::gradient

### PR DESCRIPTION
**Pull request by akva2**
returns the gradient of the function

Sits on top of https://github.com/OPM/IFEM/pull/638

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 please

File: src/Utility/Function.h [hunk 1]
```diff
@@ -251,6 +251,21 @@ class VecFunc : public utl::SpatialFunction<Vec3>, public FunctionBase
...
+    result = this->evalGradient(X);
+    return result;
+  }
+
+protected:
+  //! \brief Returns the gradient of the function as a 1D array.
+  virtual std::vector<Real> evalGradient(const Vec3& X) const
+  {
+    return {};
+  }
```
Comment (kmokstad) [inline src/Utility/Function.h:268]: The argument `X` is not used in this dummy implementation, and I prefer one-liner implementation:
`virtual std::vector<Real> evalGradient(const Vec3&) const { return {}; }`

Comment (kmokstad) [reply src/Utility/Function.h:268]: Another point, couldn't this implementation be placed in the parent class `SpatialFunction` instead? So you don't have to repeat it for each sub-class. Or are there some template issues that makes that difficult.

File: src/Utility/TensorFunction.h [hunk 2]
```diff
@@ -49,6 +50,23 @@ class TensorFunc : public utl::SpatialFunction<Tensor>, public FunctionBase
   {
     return this->evaluate(X).trace();
   }
+
+  //! \brief Evaluates first derivatives of the function.
+  utl::matrix3d<Real> gradient(const Vec3& X) const
+  {
+    const size_t nsd = sqrt(this->dim());
```
Comment (kmokstad) [inline src/Utility/TensorFunction.h:57]: `dim()` just returns the protected member `ncmp` so you could just write `sqrt(ncmp)` here.

File: src/Utility/TensorFunction.h [hunk 3]
```diff
@@ -49,6 +50,23 @@ class TensorFunc : public utl::SpatialFunction<Tensor>, public FunctionBase
...
+    utl::matrix3d<Real> result(nsd, nsd, nsd);
+    result.fill(this->evalGradient(X).data());
+
```
Comment (kmokstad) [inline src/Utility/TensorFunction.h:60]: remove this blank line

File: src/Utility/TensorFunction.h [hunk 4]
```diff
@@ -85,6 +117,28 @@ class STensorFunc : public utl::SpatialFunction<SymmTensor>,
...
+    return result;
+  }
+
+protected:
+  //! \brief Returns the gradient of the function as a 1D array.
+  virtual std::vector<Real> evalGradient(const Vec3& X) const
+  {
+    return {};
+  }
```
Comment (kmokstad) [inline src/Utility/TensorFunction.h:141]: Se previous comments on the `evalGradient()` method.

File: src/Utility/TensorFunction.h [hunk 5]
```diff
@@ -49,6 +50,23 @@ class TensorFunc : public utl::SpatialFunction<Tensor>, public FunctionBase
...
```
Comment (kmokstad) [inline src/Utility/TensorFunction.h:69]: See two comments for VecFunc which applies here as well.

**Review by kmokstad — APPROVED** [review]
I think also we should introduce the TensorFunction.C file containing the non-trival member function implementasjons.

File: src/Utility/Function.h (hunk 1 above)
Comment (akva2) [reply src/Utility/Function.h]: Ok and yeah, I initial was working towards removing the SpatialFunction layer but ended up not doing it, so I have moved the slot into that thus dropping the duplicates of the dummy.

File: src/Utility/TensorFunction.h (hunk 2 above)
Comment (akva2) [reply src/Utility/TensorFunction.h]: ok.

File: src/Utility/TensorFunction.h (hunk 3 above)
Comment (akva2) [reply src/Utility/TensorFunction.h]: ok, other projects wants blank before return so I mix things up sometimes.

**Comment by akva2** [comment]
added TensorFunction.C

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Pull request merged**